### PR TITLE
Add middleware plugins from fl4re/sinopia

### PR DESCRIPTION
**Pull request by HCanber**
This is basically PR fl4re/sinopia#18 by @fredr

While waiting for a new plugin system (#169), we need to allow middleware plugins.
Some sinopia plugins needs this, as mentioned in #353

### CODE REVIEW

**Comment by juanpicado** [comment]
@Meeeeow @ayusharma would be great if your provide here approval. This PR deserves a common approval eye 👍  Thanks @HCanber awesome.

File: src/api/index.js [hunk 1]
```diff
@@ -80,6 +81,17 @@ module.exports = function(config_hash) {
       });
     });
   }
+  // register middleware plugins
+  const plugin_params = {
+    config: config,
+    logger: Logger.logger,
+  };
+  const plugins = load_plugins(config, config.middlewares, plugin_params, function(p) {
```
Comment (juanpicado) [inline src/api/index.js]: I'd guess `p` means plugin. I think a better naming here is possible.

File: src/api/index.js [hunk 2]
```diff
@@ -80,6 +81,17 @@ module.exports = function(config_hash) {
...
+    return p.register_middlewares;
+  });
+  plugins.forEach(function(p) {
+    p.register_middlewares(app, auth, storage);
```
Comment (juanpicado) [inline src/api/index.js]: The same here.

File: src/api/index.js (hunk 1 above)
Comment (HCanber) [reply src/api/index.js]: @juanpicado Both these have been fixed and a new commit have been pushed. Let me know if you want me to squash the commits.

File: conf/full.yaml [hunk 3]
```diff
@@ -186,3 +186,8 @@ notify:
     content: ' {{ handlebar-expression }}'
     # For Slack, follow the following format:
     # content: '{ "text": "Package *{{ name }}* published to version *{{ dist-tags.latest }}*", "username": "Verdaccio", "icon_emoji": ":package:" }'
+
+# Configure plugins that can register custom middlewares
+#middlewares:
+# plugin-name:
+#   setting: true
```
Comment (Meeeeow) [inline conf/full.yaml]: setting: true => i-am-param: true

File: src/api/index.js [hunk 4]
```diff
@@ -11,6 +11,7 @@ const Cats = require('../lib/status-cats');
 const Storage = require('../lib/storage');
 const _ = require('lodash');
 const cors = require('cors');
+const load_plugins = require('../lib/plugin-loader').load_plugins;
```
Comment (Meeeeow) [inline src/api/index.js]: Please use camelCase in new code

**Comment by juanpicado** [comment]
ping @HCanber

**Pull request merged**